controllers/topology/current_state.go

```diff
@@ -18,12 +18,140 @@ package topology
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *v1alpha4.Cluster, class v1alpha4.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getTemplate(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.ControlPlaneRef.Kind, cluster.Spec.ControlPlaneRef.Name)
+	}
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the ClusterClass of the given Cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	infra, err := r.getTemplate(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries for all MachineDeployments and filters them for their linked Cluster and
+// whether they are managed by a ClusterClass using labels. A Cluster may have zero or more MachineDeployments. Zero is
+// expected on first reconcile. If MachineDeployments are found for the Cluster their Infrastructure and Bootstrap references
+// are inspected. Where these are not found the function will throw an error.
+func (r *ClusterReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *v1alpha4.Cluster) (map[string]machineDeploymentTopologyState, error) {
+	mDeploymentState := make(map[string]machineDeploymentTopologyState)
+
+	md := &v1alpha4.MachineDeploymentList{}
+	err := r.Client.List(ctx, md, client.MatchingLabels{
+		v1alpha4.ClusterLabelName:         cluster.Name,
+		v1alpha4.ClusterTopologyLabelName: "",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read MachineDeployments for managed topology")
+	}
+	for _, m := range md.Items {
+		machineDeployment := m
+		if _, ok := m.ObjectMeta.Labels[v1alpha4.ClusterTopologyMachineDeploymentLabelName]; !ok {
+			return nil, fmt.Errorf("failed to find MachineDeploymentTopology name for %s", machineDeployment.Name)
+		}
+		mdTopologyName := m.ObjectMeta.Labels[v1alpha4.ClusterTopologyMachineDeploymentLabelName]
+		if mdTopologyName == "" {
+			return nil, fmt.Errorf("no value for MachineDeploymentTopoplogyName for %s", machineDeployment.Name)
+		}
```

Comment:
```suggestion
		mdTopologyName, ok := m.ObjectMeta.Labels[v1alpha4.ClusterTopologyMachineDeploymentLabelName]
		if !ok {
			return nil, fmt.Errorf("failed to find MachineDeploymentTopology name for %s", machineDeployment.Name)
		}
		if mdTopologyName == "" {
			return nil, fmt.Errorf("no value for MachineDeploymentTopoplogyName for %s", machineDeployment.Name)
		}
```
I think it's better to do the lookup only once.

Reply:
There's a scoping issue here I think - the if statement means the key can only be used inside its scope, so we'd have to declare a variable or restructure the code. This is the compromise I landed on.

Reply:
Exactly, that's why i moved the map lookup in front of the if statement. I think it's better then doing two lookups and is what I usually see in other Go code.

Reply:
Sure - I can change it.

Reply:
Absolutely fine if you want to wait/ask for other opinions. It's just the way I personally prefer it.